File: spikes/wasm-mat4/tree.mjs
// Tree matrix multiply. world[i] = world[parent[i]] * local[i].
//
// Both kernels are inlined as base64, so there is no fetch, no bundler plugin
// and no async anywhere in the API. Run ./embed.mjs after ./build.sh to refresh
// them, and ./embed.mjs --check to verify they match the committed .wasm files.

const WASM =
    'AGFzbQEAAAABCAFgBH9/f38AAwIBAAUGAQEQgIAEBggBfwFBgIgECwcRAgZtZW1vcnkCAAR0cmVlAAAKzAIByQICA38F' +
    'e0EAIQQgA0EAIANBAEobIQUgACEGAkADQCAEIAVGDQECQAJAIAIgBEECdGooAgAiAyAESQ0AQQAhAwNAIANBwABGDQIg' +
    'BiADaiABIANq/QAAAP0LAAAgA0EQaiEDDAALCyAAIANBBnRqIgP9AAAwIQcgA/0AACAhCCAD/QAAECEJIAP9AAAAIQpB' +
    'ACEDA0AgA0HAAEYNASAGIANqIAcgASADav0AAAAiCyAL/Q0MDQ4PDA0ODwwNDg8MDQ4P/eYBIAggCyAL/Q0ICQoLCAkK' +
    'CwgJCgsICQoL/eYBIAogCyAL/Q0AAQIDAAECAwABAgMAAQID/eYBIAkgCyAL/Q0EBQYHBAUGBwQFBgcEBQYH/eYB/eQB' +
    '/eQB/eQB/QsAACADQRBqIQMMAAsLIAZBwABqIQYgAUHAAGohASAEQQFqIQQMAAsLCw==';

const WASM_FMA =
    'AGFzbQEAAAABCAFgBH9/f38AAwIBAAUGAQEQgIAEBggBfwFBgIgECwcRAgZtZW1vcnkCAAR0cmVlAAAKwwIBwAICA38F' +
    'e0EAIQQgA0EAIANBAEobIQUgACEGAkADQCAEIAVGDQECQAJAIAIgBEECdGooAgAiAyAESQ0AQQAhAwNAIANBwABGDQIg' +
    'BiADaiABIANq/QAAAP0LAAAgA0EQaiEDDAALCyAAIANBBnRqIgP9AAAwIQcgA/0AACAhCCAD/QAAECEJIAP9AAAAIQpB' +
    'ACEDA0AgA0HAAEYNASAGIANqIAcgASADav0AAAAiCyAL/Q0MDQ4PDA0ODwwNDg8MDQ4PIAggCyAL/Q0ICQoLCAkKCwgJ' +
    'CgsICQoLIAkgCyAL/Q0EBQYHBAUGBwQFBgcEBQYHIAogCyAL/Q0AAQIDAAECAwABAgMAAQID/eYB/YUC/YUC/YUC/QsA' +
    'ACADQRBqIQMMAAsLIAZBwABqIQYgAUHAAGohASAEQQFqIQQMAAsLCw==';

function decode(base64) {
    if (typeof Buffer !== 'undefined') return Buffer.from(base64, 'base64');
    const bin = atob(base64);
    const bytes = new Uint8Array(bin.length);
    for (let i = 0; i < bin.length; i++) bytes[i] = bin.charCodeAt(i);
    return bytes;
}

/**
 * True when this engine can run the relaxed SIMD kernel.
 *
 * Validates the real module rather than a stand in probe, so the check tests
 * exactly the features the kernel uses and cannot drift away from them.
 */
export function fmaSupported() {
    return WebAssembly.validate(decode(WASM_FMA));
}

/**
 * Allocates a tree of `capacity` nodes resident in wasm memory.
 *
 * Pass `fma` to use the relaxed SIMD kernel. Engines are free to fuse or not, so
 * results differ slightly between them. Leave it off when results have to match
 * across devices, as in a networked simulation or replay.
 *
 * Memory is sized once here and never grows afterwards, because `memory.grow`
 * detaches every view already handed out.
 */
export function createTree(capacity, options) {
    if (!Number.isInteger(capacity) || capacity < 0)
        throw new RangeError(`capacity must be a non negative integer, got ${capacity}`);
    // wasm32 addresses 4GiB, and a node costs 132 bytes across the three buffers
    const max = Math.floor((4294967296 - 65536) / 132);
    if (capacity > max) throw new RangeError(`capacity ${capacity} exceeds the ${max} a 4GiB wasm memory holds`);

    const fma = options?.fma === true;
    if (fma && !fmaSupported()) throw new Error('relaxed simd is unavailable on this engine');

    const instance = new WebAssembly.Instance(
        // sync compile, legal on the main thread while the module stays under 4KB
        new WebAssembly.Module(decode(fma ? WASM_FMA : WASM)),
        {},
    );
    const memory = instance.exports.memory;

    // one grow at creation, before any view exists
    const needed = Math.ceil((capacity * 132 + 65536) / 65536);
    const have = memory.buffer.byteLength / 65536;
    if (needed > have && memory.grow(needed - have) < 0)
        throw new RangeError(`could not reserve ${needed} pages of wasm memory for ${capacity} nodes`);

    const matrices = capacity * 16;
    const tree = {
        capacity,
        count: capacity,
        memory,
        fma,
        local: new Float32Array(memory.buffer, 0, matrices),
        world: new Float32Array(memory.buffer, matrices * 4, matrices),
        parent: new Int32Array(memory.buffer, matrices * 8, capacity),
    };

    /**
     * Propagates local to world for the first `count` nodes. Returns `tree`.
     *
     * A parent index outside `0 <= parent[i] < i` is treated as a root, so a
     * malformed tree gives wrong values rather than reading out of bounds. Use
     * `validate` to find such an index.
     */
    tree.update = (count = tree.count) => {
        if (!Number.isInteger(count) || count < 0 || count > tree.capacity)
            throw new RangeError(`count must be an integer in [0, ${tree.capacity}], got ${count}`);
        instance.exports.tree(tree.world.byteOffset, tree.local.byteOffset, tree.parent.byteOffset, count);
        return tree;
    };

    /**
     * Checks that every node is a root or points backwards, which is what the
     * kernel needs to resolve the tree in one pass. Returns the first index
     * failing that, or -1. Not for hot paths.
     */
    tree.validate = (count = tree.count) => {
        if (!Number.isInteger(count) || count < 0 || count > tree.capacity)
            throw new RangeError(`count must be an integer in [0, ${tree.capacity}], got ${count}`);
        for (let i = 0; i < count; i++) {
            const p = tree.parent[i];
            if (p >= 0 && p >= i) return i;
        }
        return -1;
    };

    return tree;
}

File: spikes/wasm-mat4/tree.c
#include <wasm_simd128.h>

#define EXPORT __attribute__((visibility("default")))

// broadcast lane l of v across all four lanes
#define BC(v, l) wasm_i32x4_shuffle(v, v, l, l, l, l)

// Relaxed madd fuses the multiply and add into a single operation. Engines are
// free to fuse or not, so the same input gives slightly different results on
// different engines, which is why this build is opt in.
#ifdef USE_FMA
#define MADD(acc, x, y) __builtin_wasm_relaxed_madd_f32x4(x, y, acc)
#else
#define MADD(acc, x, y) wasm_f32x4_add(acc, wasm_f32x4_mul(x, y))
#endif

/*
 * world[i] = world[parent[i]] * local[i], with a negative parent meaning root.
 *
 * Parents must precede children, so one forward pass resolves the whole tree.
 * An index that does not satisfy 0 <= parent[i] < i is treated as a root, which
 * keeps every read inside the buffer whatever the caller passes.
 * Column major, so out column j is the sum over k of parent column k scaled by
 * local[j * 4 + k].
 *
 * world may not overlap local or parent. Reading world[p] while writing
 * world[i] is fine, since p is always less than i.
 */
EXPORT void tree(float *world, const float *local, const int *parent, int n) {
    for (int i = 0; i < n; i++) {
        unsigned p = (unsigned)parent[i];
        const float *L = local + i * 16;
        float *W = world + i * 16;

        // one unsigned compare rejects a root (negative), a forward reference and
        // an out of range index at once. Anything invalid is treated as a root,
        // so p < i < n always holds below and the read cannot leave the buffer.
        if (p >= (unsigned)i) {
            for (int j = 0; j < 4; j++) wasm_v128_store(W + j * 4, wasm_v128_load(L + j * 4));
            continue;
        }

        const float *P = world + p * 16;
        v128_t a0 = wasm_v128_load(P);
        v128_t a1 = wasm_v128_load(P + 4);
        v128_t a2 = wasm_v128_load(P + 8);
        v128_t a3 = wasm_v128_load(P + 12);

        for (int j = 0; j < 4; j++) {
            v128_t bj = wasm_v128_load(L + j * 4);
            v128_t r = wasm_f32x4_mul(a0, BC(bj, 0));
            r = MADD(r, a1, BC(bj, 1));
            r = MADD(r, a2, BC(bj, 2));
            r = MADD(r, a3, BC(bj, 3));
            wasm_v128_store(W + j * 4, r);
        }
    }
}
